ArduCopter/precision_landing.cpp

```diff
@@ -22,4 +27,201 @@ void Copter::update_precland()
+                // We eventually want to retry landing, but lets descend for some time and hope the target gets in sight
+                // If not, we will retry landing
+                landing_target_lost_action = TargetLostAction::DESCEND;
+                break;
+            case AC_PrecLand::PldRetryStrictness::NOT_STRICT:
+                // User just wants to land, prec land isn't a priority
+                landing_target_lost_action = TargetLostAction::LAND_VERTICALLY;
+                break;
+        }
+        }
+        break;
+
+    case TargetLostAction::DESCEND:
+        copter.flightmode->run_land_controllers();
+        if (AP_HAL::millis() - copter.precland.get_last_precland_output_ms() >= copter.precland.get_min_retry_time_sec()) {
+            // we have descended for some time and the target still isn't in sight
+            // lets retry
+            landing_target_lost_action = TargetLostAction::RETRY_LANDING;
+            retry_state = RetryLanding::INIT;
+        }
+        break;
+
+    case TargetLostAction::RETRY_LANDING:
+        // retry the landing by going to the last known horizontal position of the target
+        retry_landing();
+        break;
+
+    case TargetLostAction::LAND_VERTICALLY:
+        // Just land vertically
+        copter.flightmode->run_land_controllers();
+        break;
+    }
+}
+
+// Retry landing based on a previously known location of the landing target
+void Mode::PrecLand_StateMachine::retry_landing()
+{
+    if (_retry_count > copter.precland.get_max_retry_allowed() || copter.precland.get_max_retry_allowed() == 0) {
+        // we have exhausted the amount of times vehicle was allowed to retry landing
+        // do failsafe measure so the vehicle isn't stuck in a constant loop
+        do_failsafe_actions();
+        return;
+    }
+
+    // get the last known location of the landing target
+    Vector3f last_known_loc;
+    copter.precland.get_last_detected_landing_location(last_known_loc);
+
+    switch (retry_state) {
+    case RetryLanding::INIT:
+        // Init the Retry
+        _retry_count ++;
+        copter.flightmode->land_retry_position(last_known_loc);
+        // initialise the position controller
+        copter.pos_control->init_z_controller_no_descent();
+        copter.pos_control->init_xy_controller();
+        retry_state = RetryLanding::IN_PROGRESS;
+        // inform the user what we are doing
+        copter.gcs().send_text(MAV_SEVERITY_INFO, "Retrying Precision Landing");
+        break;
+
+    case RetryLanding::IN_PROGRESS:
+    {
+        // continue converging towards the target till we are close by
+        copter.flightmode->land_retry_position(last_known_loc);
+        const float dist_to_target_xy = copter.pos_control->get_pos_error_xy_cm();
+        const float dist_to_target_z = copter.pos_control->get_pos_error_z_cm();
+        copter.gcs().send_text(MAV_SEVERITY_CRITICAL, "hello world! %5.3f", (double)dist_to_target_z * 0.01f);
```

Comment:
hello world ?

Reply:
A poor case of debug code pushed to commit. My bad!